### PR TITLE
Add MorphemeList.splittedWith method

### PR DESCRIPTION
**Pull request by eiennohito**
1. Make all tokenizing methods return concrete MorphemeList instead of List<Morpheme>
2. Add new `MorphemeList.split` API which resplits already-splitted MorphemeList with different analysis mode. Making more fine analysis from more coarse analysis is supported, going back is not.
3. Refactor splitting logic to remove duplicated code.

If there are better ideas on the name of this method, I would be happy to hear them.

### CODE REVIEW

File: src/main/java/com/worksap/nlp/sudachi/LatticeNodeImpl.java [hunk 1]
```diff
@@ -134,4 +136,30 @@ public String toString() {
...
+        } else if (mode == Tokenizer.SplitMode.B) {
+            appendSplitted(result, getWordInfo().getBunitSplit());
+        } else {
+            result.add(this);
+        }
+    }
+
+    void appendSplitted(List<LatticeNode> result, int[] ids) {
+        if (ids.length <= 1) {
+            result.add(this);
```
Comment (kazuma-t) [inline src/main/java/com/worksap/nlp/sudachi/LatticeNodeImpl.java:153]: In the original code, it was obvious to add `this`, but since this part was cut out, even if the length of `ids` is 1, the content is not necessarily `this`.

File: src/test/java/com/worksap/nlp/sudachi/JapaneseTokenizerTest.java [hunk 2]
```diff
@@ -256,6 +257,46 @@ public void disableEmptyMorpheme() throws IOException {
...
+    }
+
+    @Test
+    public void splitAfterTokenizeBtoC() {
+        MorphemeList morphemes1 = tokenizer.tokenize(Tokenizer.SplitMode.B, "東京都");
+        assertThat(morphemes1.size(), is(1));
+        MorphemeList morphemes2 = morphemes1.splittedWith(Tokenizer.SplitMode.C);
+        assertThat(morphemes2, sameInstance(morphemes1));
+    }
+
```
Comment (kazuma-t) [inline src/test/java/com/worksap/nlp/sudachi/JapaneseTokenizerTest.java:321]: To be sure, please also run a test that produces a zero-length morpheme in the middle.

**Review by kazuma-t — CHANGES_REQUESTED** [review]

File: src/test/java/com/worksap/nlp/sudachi/JapaneseTokenizerTest.java (hunk 2 above)
Comment (eiennohito) [reply src/test/java/com/worksap/nlp/sudachi/JapaneseTokenizerTest.java:321]: Will do.

File: src/main/java/com/worksap/nlp/sudachi/LatticeNodeImpl.java (hunk 1 above)
Comment (eiennohito) [reply src/main/java/com/worksap/nlp/sudachi/LatticeNodeImpl.java:153]: I do not understand this comment.

0-length splits means that the current token is not splittable and should be added as it is.
1-length splits mean... that the current token should be replaced with another one? Is this behavior actually used?

Comment (kazuma-t) [reply src/main/java/com/worksap/nlp/sudachi/LatticeNodeImpl.java:153]: It is possible to put something different from `this` in the `ids`, and while we don't usually do that, if we do, it would be an incomprehensible bug, so it should be checked.

Comment (eiennohito) [reply src/main/java/com/worksap/nlp/sudachi/LatticeNodeImpl.java:153]: This function is not public and will could not be called outside sudachi (except with reflection).

Comment (kazuma-t) [reply src/main/java/com/worksap/nlp/sudachi/LatticeNodeImpl.java:153]: When someone only looks at the signature of this method, they would not expect it to put `this` in if the length of the `ids` was 1, regardless of the contents of the `ids`.

Comment (eiennohito) [reply src/main/java/com/worksap/nlp/sudachi/LatticeNodeImpl.java:153]: Handling 1-length case now as rewriting of the morpheme if it has a different word id.

Comment (kazuma-t) [reply src/main/java/com/worksap/nlp/sudachi/LatticeNodeImpl.java:153]: Thank you!

**Review by kazuma-t — APPROVED** [review]
